File: lib-universalAdapter/src/main/java/com/llj/adapter/observable/ListObserverListener.java
package com.llj.adapter.observable;


import androidx.annotation.Nullable;

/**
 * Interface which listens to {@link ListObserver} changes.
 *
 * @param <T> The type of items stored in the observed list.
 */
public interface ListObserverListener<T> {

  void onItemRangeChanged(ListObserver<T> observer, int start, int count, @Nullable Object payload);

  void onItemRangeChanged(ListObserver<T> observer, int start, int count);

  void onItemRangeInserted(ListObserver<T> observer, int start, int count);

  void onItemRangeRemoved(ListObserver<T> observer, int start, int count);

  void onGenericChange(ListObserver<T> observer);
}

File: lib-universalAdapter/src/main/java/com/llj/adapter/observable/ListObserverListener.kt
package com.llj.adapter.observable

/**
 * Interface which listens to [ListObserver] changes.
 *
 * @param <T> The type of items stored in the observed list.
 */
interface ListObserverListener<out T> {

    fun onItemRangeChanged(observer: ListObserver<@UnsafeVariance T>, startPosition: Int, count: Int, payload: Any?)
    fun onItemRangeChanged(observer: ListObserver<@UnsafeVariance T>, startPosition: Int, count: Int)
    fun onItemRangeInserted(observer: ListObserver<@UnsafeVariance T>, startPosition: Int, count: Int)
    fun onItemRangeRemoved(observer: ListObserver<@UnsafeVariance T>, startPosition: Int, count: Int)
    fun onGenericChange(observer: ListObserver<@UnsafeVariance T>)
}